Update the distributed sequencer docs ready for V1 (#1269)
LFDT-Paladin/paladin: doc-site/docs/architecture/distributed_sequencer_architecture.md

The distributed sequencer is one of those standalone components and the sequencer manager handles the lifetime of sequencers for different domain contracts.

> The distributed sequencer is implemented using [state machines](./distributed_sequencer_state_machine.md) to track activity in the [2 core components](./distributed_sequencer_components.md) of the sequencer (the originator and the coordinator).

Reviewer: Links to components which doesn't exist.

I also found it jarring having read about these 2 core components to then read about the 3 key components.

For every active sequencer the state machines are updated by generating, receiving (from other nodes), and processing state machine events. The order that state machine events are handled is critical for reliable transaction processing. The sequencer architecture uses several event handling queues, implemented using Golang channels, to maintain this ordering.

## Sequencer Components

In domains (such as Pente) where the spending rules for states allow any one of a group of parties to spend the state, then we need to coordinate the assembly of transactions across multiple nodes so that we can maximize the throughput by speculatively spending new states and avoiding transactions being reverted due to double concurrent spending / state contention.

To achieve this, it is important that we have an algorithm that allows all nodes to agree on which of them should be selected as the coordinator at any given point in time. And all other nodes delegate their transactions to the coordinator.

A node uses 3 key components to coordinate transactions with a domain contract:

> ### 1 - Sequencer

Reviewer: I would have said that its the originator who manages the overall lifecycle of transactions submitted to _this_ node - delegation, green light to approve dispatch, resuming etc

The sequencer manages the overall lifecycle of transactions submitted to the node. The sequencer comprises 2 sub-components:

### 2 - Originator

The originator is responsible for assembling and transactions when instructed to do so by the coordinator.

### 3 - Coordinator

The coordinator determines which contract-wide states should be spent in order to satisfy a transaction's inputs and communicates with originators to instruct them what to submit to the EVM.

> A coordinator may not always be running on every node participating in the private contract (see below).

Reviewer: not active rather than not running?

---
classDiagram
    Sequencer "1" --> "1" Originator
    Sequencer "1" --> "1" Coordinator
    Originator "1" --> "*"  OriginatorTransaction : Transactions
    Coordinator "1" --> "*"  CoordinatorTransaction : Transactions
    class Sequencer{
        Address ContractAddress
    }
    class Originator {
    }
    class Coordinator {
    }
    class OriginatorTransaction {
    }
    class CoordinatorTransaction {
    }
```

## Sequencer lifecycle

Paladin runs a dedicated sequencer for every domain contract it is actively processing transactions for. The sequencers used for different contracts have specific committee members, originator nodes, and coordination behaviours.

Some contracts will be single-use (or very infrequent use) where others may regularly have new transactions being submitted. Paladin manages the lifecycle of every sequencer to avoid runtime resources being used if no transactions are actively being handled for a given domain instance.

The seqencer lifecycle management built in to Paladin is highly configurable, but the default configuration is intended to provide:

- minimal resource consumption for infrequently used contracts

>   - a sequencer will be removed from memory whe not in use

Reviewer: ```suggestion
  - a sequencer will be removed from memory when not in use
```

These indented bullets need an extra space (or two) to be correctly rendered - it's appearing as a flat list of bullets right now

- minimal latency for frequently used contracts
  - a sequencer will not be removed from memory if it is actively processing transactions
- stability of the Paladin runtime by limited the overall number of sequencers which are active at a given time

The following sequence diagram shows the lifecycle of a sequencer for a newly submitted transaction:

> ![Sequencer Lifecycle](./diagrams/paladin-sequencer-tx-flow.svg){.zoomable-image}

Reviewer: A few observations on this diagram
* `State Assembling-> Endorsing -> Dispatching` (transactions states) is put at the same level as `State Idle -> Active` (coordinator states) - I think this is supposed to be the lifecycle of the sequencer rather than the transaction?
* Endorsement/Predispatch request is not shown (probably intentionally?)
* nonce assigned line is going from sequencer manager to coordinator - this isn't accurate
* heartbeat interval is 10s by default not 500ms

## Event driven state machines

> The distributed sequencer is built using [state machines](./distributed_sequencer_state_machine.md) to track the coordinator, originator, and their transactions. A single Go route called the `event loop` processes all events for the coordinator and originator respectively. This ensures thread safe updating of the state machines.

Reviewer: ```suggestion
The distributed sequencer is built using [state machines](./distributed_sequencer_state_machine.md) to track the coordinator, originator, and their transactions. A single goroutine called the `event loop` processes all events for the coordinator and originator respectively. This ensures thread safe updating of the state machines.
```

This reads like its one goroutine for both coordinator and originator, the `respectively` doesn't really capture that its one each.

The state machines are updated by the event loop processing events. Events typically result in a transition between the states and in some cases result in specific actions being carried out before or during the transition.

> Events can be received from sources external to the node (such as another node), or internally from a state machine produce events as a result of processing other ones.

Reviewer: ```suggestion
Events can be received from sources external to the node (such as another node), or internally from a state machine producing events as a result of processing other ones.
```

The following diagram demonstrates how events are passed to the state machines and how events are processed in one of two priority categories. Priority events are always processed first by the relevant state machine, while regular events are processed whenever there are no priority events to handle:

> ![Event Handling](diagrams/paladin-code-architecture.svg){.zoomable-image}

Reviewer: The handle event box should probably be between the coord state machine loop and coord tx state machines diagrams

Reviewer: dependent TX events are within a domain instance, not cross domain instance
